docs/design_notes/setup_or_nncompact.rst

```diff
@@ -0,0 +1,78 @@
+Should I use ``setup`` or ``nn.compact``?
+=========================================
+
+In Flax's module system (named `Linen`_), submodules, parameters and/or other variables
+can be defined in two ways:
+
+1. **Explicitly** (using ``setup``):
+
+   Assign to ``self.<attr>`` inside a :meth:`setup <flax.linen.Module.setup>`
+   method. Then read ``self.<attr>`` from any "forward pass" method defined on the class.
+   This resembles how most objects work in Python, and how modules are defined in PyTorch.
+2. **In-line** (using ``nn.compact``):
+  
+   Declare directly within a single "forward pass" method annotated
+   with :meth:`nn.compact <flax.linen.compact>`. This allows you to define your whole module
+   in a single method, and "co-locate" submodules, parameters and variables next to 
+   where they are used.
+
+**Both of these approaches are perfectly valid, behave the same way, and interoperate with all of Flax**.
+
+Here is a short example of a module defined in both ways. 
```

Comment:
I would also explain how naming of submodules works for `setup`

Reply:
yes, good point. I'll add something about that.